src/charm.py

```diff
@@ -106,34 +106,42 @@ def _on_install(self, event: InstallEvent):
         channel = self.model.config["snap-channel"]
         self._snap.install(channel=channel, core_res=core_res, nats_res=nats_res)
 
-    def _on_config_changed(self, event: ConfigChangedEvent):
-        self._reconfigure_nats(self.model.config)
+    def _on_upgrade_charm(self, event: UpgradeCharmEvent):
+        self.ca_client.restore_state()
 
-    def _on_tls_config_ready(self, event):
+    def _on_config_changed(self, event: ConfigChangedEvent):
+        config = dict(self.model.config)
+        if (
+            not (config["tls-cert"] or config["tls-key"] or config["tls-ca-cert"])
+            and self.ca_client.is_ready
+        ):
+            logger.info("Configuring CA certificates from relation")
+            key = self.ca_client.key.private_bytes(
```

Comment:
Is there any chance that self.ca_client.key be None if the internal state for the ca client was not set E.g self.state.key == None. Same for ca_certificate/certificate

Reply:
The condition above this has a check for nullability. The `self.ca_client.is_ready` attribute checks for the existence of all 3 properties.

Reply:
Okay, that makes sense to me.